### PR TITLE
planner: LRUPlanCache memory trace

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #37632

Problem Summary:
Count the memory usage of LRUPlanCache

### What is changed and how it works?
Implement MemoryUsage() for 
* planCacheKey
* PlanCacheValue
* LRUPlanCache
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_lru.go [hunk 1]
```diff
@@ -200,6 +202,32 @@ func (l *LRUPlanCache) memoryControl() {
 	}
 }
 
+const emptyLRUPlanCacheSize = int64(unsafe.Sizeof(LRUPlanCache{}))
+
+// MemoryUsage return the memory usage of LRUPlanCache
+func (l *LRUPlanCache) MemoryUsage() (sum int64) {
```
Comment (qw4990) [inline planner/core/plan_cache_lru.go:208]: It seems like we should maintain its memory usage dynamically in `l.Put`.

File: planner/core/plan_cache_utils.go [hunk 2]
```diff
@@ -325,12 +341,45 @@ type PlanCacheValue struct {
...
+	case PhysicalPlan:
+		sum = x.MemoryUsage()
+	case *Insert:
+		sum = x.MemoryUsage()
+	case *Update:
+		sum = x.MemoryUsage()
+	case *Delete:
+		sum = x.MemoryUsage()
+	default:
+		logutil.BgLogger().Info("this Plan did`t implement MemoryUsage method")
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:368]: How about removing this log and just using `sum=SomeDefaultValue`?

File: planner/core/plan_cache_utils.go [hunk 3]
```diff
@@ -325,12 +341,45 @@ type PlanCacheValue struct {
...
 
 func (v *PlanCacheValue) varTypesUnchanged(txtVarTps []*types.FieldType) bool {
 	return v.ParamTypes.CheckTypesCompatibility4PC(txtVarTps)
 }
 
+// MemoryUsage return the memory usage of PlanCacheValue
+func (v *PlanCacheValue) MemoryUsage() (sum int64) {
+	if v == nil {
+		return
+	}
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:356]: If it's initialized, we can just return the PlanCacheValueMem:
```
if v.PlanCacheValueMem > 0 {
    return v.PlanCacheValueMem
}
```

File: planner/core/plan_cache_utils.go [hunk 4]
```diff
@@ -325,12 +341,45 @@ type PlanCacheValue struct {
...
+	sum += size.SizeOfInterface + size.SizeOfSlice*2 + int64(cap(v.OutPutNames)+cap(v.ParamTypes))*size.SizeOfPointer +
+		size.SizeOfMap + int64(len(v.TblInfo2UnionScan))*(size.SizeOfPointer+size.SizeOfBool) + size.SizeOfInt64*2
+
+	for _, name := range v.OutPutNames {
+		sum += name.MemoryUsage()
+	}
+	for _, ft := range v.ParamTypes {
+		sum += ft.MemoryUsage()
+	}
+	return
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:380]: Better to cache the value:
```
v.PlanCacheValueMem = sum
```

File: planner/core/plan_cache_utils.go [hunk 5]
```diff
@@ -325,12 +341,45 @@ type PlanCacheValue struct {
 	OutPutNames       []*types.FieldName
 	TblInfo2UnionScan map[*model.TableInfo]bool
 	ParamTypes        FieldSlice
+	PlanCacheKeyMem   int64
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:344]: Better to move this filed to `planCacheKey`.

File: planner/core/plan_cache_lru.go [hunk 6]
```diff
@@ -67,23 +73,28 @@ func NewLRUPlanCache(capacity uint, guard float64, quota uint64,
 		pickFromBucket: pickFromBucket,
 		quota:          quota,
 		guard:          guard,
+		memTracker:     newTrackerForLRUPC(),
 	}
 }
 
 // strHashKey control deep or Shallow copy of string
-func strHashKey(key kvcache.Key, deepCopy bool) string {
+func strHashKey(key kvcache.Key, deepCopy bool) (string, int64) {
```
Comment (qw4990) [inline planner/core/plan_cache_lru.go:81]: How about just returning `string` here?

**Review by Reminiscent — APPROVED** [review]

File: planner/core/plan_cache_utils.go [hunk 7]
```diff
@@ -325,12 +345,52 @@ type PlanCacheValue struct {
...
+	memoryUsage       int64
 }
 
 func (v *PlanCacheValue) varTypesUnchanged(txtVarTps []*types.FieldType) bool {
 	return v.ParamTypes.CheckTypesCompatibility4PC(txtVarTps)
 }
 
+// unKnownMemoryUsage represent the memory usage of uncounted structure, maybe need implement later
+// 7 has no special meaning
+const unKnownMemoryUsage = 7
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:357]: Seems too small. Actually, I have tested it before and got an empirical value `100KB`, you can see it in our doc https://docs.pingcap.com/tidb/dev/sql-prepared-plan-cache#memory-management-of-prepared-plan-cache

File: planner/core/plan_cache_lru.go [hunk 8]
```diff
@@ -222,3 +236,24 @@ func PickPlanFromBucket(bucket map[*list.Element]struct{}, paramTypes []*types.F
...
 	return nil, false
 }
+
+const emptyLRUPlanCacheSize = int64(unsafe.Sizeof(LRUPlanCache{}))
+
+// newTrackerForLRUPC return a tracker which consumed emptyLRUPlanCacheSize
+// todo: pass label when track general plan cache memory
+func newTrackerForLRUPC() *memory.Tracker {
+	m := memory.NewTracker(memory.LabelForPreparedPlanCache, -1)
+	m.Consume(emptyLRUPlanCacheSize)
```
Comment (qw4990) [inline planner/core/plan_cache_lru.go:246]: How about ignoring this part of memory usage?

File: planner/core/plan_cache_lru.go [hunk 9]
```diff
@@ -222,3 +236,24 @@ func PickPlanFromBucket(bucket map[*list.Element]struct{}, paramTypes []*types.F
...
+
+// elementMemoryUsage return the sum of planCacheKey and planCacheValue
+func elementMemoryUsage(e *list.Element) (sum int64) {
+	pVal, ok1 := e.Value.(*planCacheEntry).PlanValue.(*PlanCacheValue)
+	pKey, ok2 := e.Value.(*planCacheEntry).PlanValue.(*planCacheKey)
+	if !ok1 || !ok2 {
+		return
+	}
+	return pKey.MemoryUsage() + pVal.MemoryUsage()
+}
```
Comment (qw4990) [inline planner/core/plan_cache_lru.go:259]: Please implement this function as `planCacheEntry.MemoryUsage`

File: planner/core/plan_cache_lru_test.go [hunk 10]
```diff
@@ -361,3 +363,41 @@ func TestIssue38244(t *testing.T) {
...
+	var res int64 = 0
+
+	// put
+	plans := []Plan{&Insert{}, &Update{}, &Delete{}}
+	for i, p := range plans {
+		k := &planCacheKey{database: strconv.FormatInt(int64(i), 10)}
+		v := &PlanCacheValue{Plan: p}
+		lru.Put(k, v, pTypes)
+		res += k.MemoryUsage() + v.MemoryUsage()
+		fmt.Println(lru.MemoryUsage(), res)
```
Comment (qw4990) [inline planner/core/plan_cache_lru_test.go:383]: Should remove this line.

**Review by qw4990 — APPROVED** [review]

**Pull request merged**